File: designPatter/Strategy/examples/ducks/C++/Duck.cpp
#include <iostream>
#include "Duck.h"
#include "FlyBehavior.h"
#include "QuackBehavior.h"

using namespace std;

Duck::Duck()
{}

void Duck::swin()
{
	cout << "Eu estou nadando!!!! " << endl;
}

void Duck::performQuack()
{
	this->quackBehavior->quack();
}

void Duck::performFly()
{
	this->flyBehavior->fly();
}

void Duck::setFlyBehavior(FlyBehavior *fl)
{
	this->flyBehavior = fl;
}

void Duck::setQuackBehavior(QuackBehavior *qc)
{
	this->quackBehavior = qc;
}

File: designPatter/Strategy/examples/ducks/ruby/duck.rb
# Abstract class that encapsulate Duck behaviour.
class Duck

  @flyBehavior
  @quackBehavior

  def swin
    raise NotImplementedError
  end

  # Introduce what kind of duck is the child.
  def display
    raise NotImplementedError
  end

  # Perform quack behaviour, independent of what kind of duck.
  def performDuck
    @quackBehavior.quack
  end

  # Perform fly behaviour, independent of what kind of duck.
  def performFly
    @flyBehavior.fly
  end

  # Set the fly behaviour.
  def setFlyBehaviour(pFlyBehaviour)
    @flyBehavior = pFlyBehaviour
  end

  def setQuackBehaviour(pQuackBehaviour)
    @quackBehavior = pQuackBehaviour
  end

end
